chips/earlgrey/src/aes.rs

```diff
@@ -318,32 +392,116 @@ impl<'a> hil::symmetric_encryption::AES128<'a> for Aes<'a> {
                 }
             }
         }
+
+        if self.deferred_call.get() {
+            panic!("Operation already in progress");
+        }
+
+        let ret;
         self.dest.replace(dest);
-        // The crypt API has two cases: separate source and destination
-        // buffers and a single source buffer.
-        // If we don't have a separate source buffer, we overwrite the
-        // destination with the data. This means that read index and write
-        // index match
-        // If we do have a separate source buffer, we start writing from
-        // 0 and the read index is separate.
         match source {
             None => {
-                self.do_crypt(start_index, stop_index, start_index);
+                ret = self.do_crypt(start_index, stop_index, start_index);
             }
             Some(src) => {
                 self.source.replace(src);
-                self.do_crypt(start_index, stop_index, 0);
+                ret = self.do_crypt(start_index, stop_index, 0);
             }
         }
-        self.client.map(|client| {
-            client.crypt_done(self.source.take(), self.dest.take().unwrap());
-        });
-        None
+
+        if ret.is_ok() {
+            // Schedule a deferred call
+            self.deferred_call.set(true);
+            self.deferred_handle
+                .map(|handle| self.deferred_caller.set(*handle));
+            None
+        } else {
+            Some((ret, self.source.take(), self.dest.take().unwrap()))
+        }
+    }
+}
+
+impl kernel::hil::symmetric_encryption::AES128Ctr for Aes<'_> {
+    fn set_mode_aes128ctr(&self, encrypting: bool) {
+        if !self.idle() {
+            panic!("AES is not idle");
+        }
+
+        self.mode.set(Mode::AES128CTR);
+
+        let mut ctrl = if encrypting {
+            CTRL::OPERATION::Encrypting
+        } else {
+            CTRL::OPERATION::Decrypting
+        };
+        ctrl += CTRL::MODE::AES_CTR;
+        // Tock only supports 128-bit keys
+        ctrl += CTRL::KEY_LEN::Key128;
+        ctrl += CTRL::MANUAL_OPERATION::CLEAR;
+
+        // We need to set the control register twice as it's shadowed
+        self.registers.ctrl.write(ctrl);
+        self.registers.ctrl.write(ctrl);
     }
 }
 
 impl kernel::hil::symmetric_encryption::AES128ECB for Aes<'_> {
     fn set_mode_aes128ecb(&self, encrypting: bool) {
-        self.configure(encrypting);
+        if !self.idle() {
+            panic!("AES is not idle");
+        }
```

Comment:
This is unfortunate.

Reply:
Yeah, we can't return an error so I'm not sure what else to do

Reply:
I'm planning on working on capsules as well, if I end up changing the HIL I can fix this up and remove these